feat: add JJOptimist's Homepage realm to examples

JJOptimist opened the pull request:
This pull request adds a new realm example to the Gno examples repository—JJOptimist's Homepage.

The realm includes:
- Personal introduction and contact information
- Configuration management system
- Owner controls 
- Tests for ownership and rendering

Gno2D2 commented:
#### 🛠 PR Checks Summary
All **Automated Checks** passed. ✅

##### Manual Checks (for Reviewers):
- [ ] **IGNORE** the bot requirements for this PR (force green CI check)
- [ ] The pull request description provides enough details


<details><summary>Read More</summary>

🤖 This bot helps streamline PR reviews by verifying automated checks and providing guidance for contributors and reviewers.

##### ✅ Automated Checks (for Contributors):
 🟢 Maintainers must be able to edit this pull request ([more info](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork))


##### ☑️ Contributor Actions:
1. Fix any issues flagged by automated checks.
2. Follow the Contributor Checklist to ensure your PR is ready for review.
	- Add new tests, or document why they are unnecessary.
	- Provide clear examples/screenshots, if necessary.
	- Update documentation, if required.
	- Ensure no breaking changes, or include `BREAKING CHANGE` notes.
	- Link related issues/PRs, where applicable.

##### ☑️ Reviewer Actions:
1. Complete manual checks for the PR, including the guidelines and additional checks if applicable.

##### 📚 Resources:
- [Report a bug with the bot](https://github.com/gnolang/gno/issues/3238).
- [View the bot’s configuration file](https://github.com/gnolang/gno/tree/master/contribs/github-bot/internal/config/config.go).

<details><summary><b>Debug</b></summary><blockquote>
<details><summary><b>Automated Checks</b></summary><blockquote>

<details><summary>Maintainers must be able to edit this pull request (more info)</summary><blockquote>

### If
```
🟢 Condition met
└── 🟢 The pull request was created from a fork (head branch repo: JJOptimist/gno)

```
### Then
```
🟢 Requirement satisfied
└── 🟢 Maintainer can modify this pull request

```
</blockquote></details>

</blockquote></details>


<details><summary><b>Manual Checks</b></summary><blockquote>

<details><summary>**IGNORE** the bot requirements for this PR (force green CI check)</summary><blockquote>

### If
```
🟢 Condition met
└── 🟢 On every pull request

```
### Can be checked by

- Any user with comment edit permission

</blockquote></details>

<details><summary>The pull request description provides enough details</summary><blockquote>

### If
```
🟢 Condition met
└── 🟢 Not (🔴 Pull request author is a member of the team: core-contributors)

```
### Can be checked by
 - team core-contributors

</blockquote></details>

</blockquote></details>

</blockquote></details>

</details>

leohhhn commented on examples/gno.land/r/jjoptimist/config.gno:
```diff
@@ -0,0 +1,51 @@
+)
+
+var (
+    
```
Unnecessary newline

leohhhn commented on examples/gno.land/r/jjoptimist/config.gno:
```diff
@@ -0,0 +1,51 @@
+)
+
+func init() {
+    owner = OWNER_ADDR
```
If you're defining this at top-level, you can simply initiate it where you declared it. 

This is a bit of a blurry line in Gno, so bear with me- following Go standards, if you can declare+initialize something immediately, do so at top level; `init` should be reserved for "doing" things you cannot do at declaration+initialization.

leohhhn commented on examples/gno.land/r/jjoptimist/config.gno:
```diff
@@ -0,0 +1,51 @@
+    
+    OWNER_ADDR = std.Address("g16vfw3r7zuz43fhky3xfsuc2hdv9tnhvlkyn0nj")
+
+    owner std.Address
```
Why not use `p/demo/ownable`?

JJOptimist replied:
Great suggestion, I've integrated `p/demo/ownable` for better ownership management in my latest changes.

leohhhn commented on examples/gno.land/r/jjoptimist/config.gno:
```diff
@@ -0,0 +1,51 @@
+    return addr == owner
+}
+
+func GetConfig() struct {
+    title       string
+    description string
+    github     string
+} {
```
Nit: I think there is no need to fold this line

leohhhn commented on examples/gno.land/r/jjoptimist/home_test.gno:
```diff
@@ -0,0 +1,70 @@
+)
+
+func TestOwnership(t *testing.T) {
+    
```
Random newline

leohhhn reviewed (changes requested):
Please check comments 🙏 

Apart from that, if you want to merge something into the `examples/` folder, it should provide some educational value about Gno to newcomers. Please, explore a bit more and see what interesting realms or packages you can use to make your home realm more unique, so that it can be merged into the monorepo. 

If you do not want to do this, you can simply deploy persmissionlessly.

leohhhn commented:
Hey @JJOptimist, what's the status of this PR?

JJOptimist commented:
> Hey @JJOptimist, what's the status of this PR?
> 
I fixed the things you pointed out and took your suggestion to use ownable. Ive been experimenting to make something creative/ worth the HoF, so i will push it all together soon

JJOptimist commented:
I've updated the realm

- Integrated `p/demo/ownable` for proper ownership management and access control
- Added dynamic ASCII art gnome that changes based on block height
- Moved files to `home/` directory structure
- Added proper tests for both ownership and configuration

The gnome's expression changes based on block height divisibility (by 2, 3, 5, or 7)

jefft0 commented on mykey_backup.armor:
Should this file be removed?

leohhhn commented on mykey_backup.armor:
Leftover file?

leohhhn commented:
@JJOptimist 

Generally, please take a look at comments and reply to them, even if they're outdated because you've changed the code completely.

JJOptimist commented:
Thanks for your patience. Let me explain the code evolution:

Initially, I tried to create an animated realm with CSS animations. However, I discovered that complex CSS animations weren't the best approach for a Gno realm, as:
   - CSS animations added unnecessary complexity
   - The focus should be on demonstrating Gno-specific features
   - Markdown rendering is more reliable for realm display

leohhhn commented on examples/gno.land/r/jjoptimist/home/home_test.gno:
```diff
@@ -0,0 +1,41 @@
+    "testing"
+)
+
+func TestOwnership(t *testing.T) {
+    owner := GetOwner()
+    if owner.String() != "g16vfw3r7zuz43fhky3xfsuc2hdv9tnhvlkyn0nj" {
+        t.Errorf("Expected owner to be g16vfw3r7zuz43fhky3xfsuc2hdv9tnhvlkyn0nj, got %s", owner.String())
+    }
+
+    std.TestSetOrigCaller(owner)
+    if !o.CallerIsOwner() {
+        t.Error("Owner address should return true for CallerIsOwner check")
+    }
+
+    nonOwner := std.Address("g1jg8mtutu9khhfwc4nxmuhcpftf0pajdhfvsqf5")
+    std.TestSetOrigCaller(nonOwner)
+    if o.CallerIsOwner() {
+        t.Error("Non-owner address should return false for CallerIsOwner check")
+    }
+}
```
You generally don't need to test functionality of another imported package; what you should do, ideally, is check the package code, and verify its tests. If they're satisfactory, you opt into using the package. 

By doing what you did in the code, you're creating boilerplate and taking up unnecessary space on the chain :)

JJOptimist replied:
I removed unnecessary tests for imported packages.

leohhhn commented on examples/gno.land/r/jjoptimist/home/home.gno:
```diff
@@ -0,0 +1,80 @@
+
+var (
+    creation time.Time
+    gnomeArt1 = `   /\
+  /  \
+ ,,,,,
+(o.o)
+(\_/)
+-"-"-`
+
+    gnomeArt2 = `   /\
+  /  \
+ ,,,,,
+(^.^)
+(\_/)
+ -"-`
+
+    gnomeArt3 = `   /\
+  /  \
+ ,,,,,
+(*.*)
+(\_/)
+"-"-"`
+
+    gnomeArt4 = `   /\
+  /  \
+ ,,,,,
+(o.~)
+(\_/)
+ -"-`
```
These can be `consts`, no?

JJOptimist replied:
Yes, first idea was for them to have animations but now that they are static it makes more sense to be `consts`.

leohhhn commented on examples/gno.land/r/jjoptimist/home/gno.mod:
```diff
@@ -0,0 +1,5 @@
+module gno.land/r/jjoptimist
+
+require (
+    gno.land/p/demo/ownable v0.0.0
+) 
```
The `require` keyword was removed a month or so ago. This will make tests & the CI fail

JJOptimist replied:
I removed it. At first, i didn't check all the tests I wrote for my realm and thought that was the thing making it fail but in the end, it was `require`.

leohhhn commented on examples/gno.land/r/jjoptimist/home/config.gno:
```diff
@@ -0,0 +1,35 @@
+}
+
+func UpdateConfig(newTitle, newDescription, newGithub string) {
+    if !o.CallerIsOwner() {
+        panic("not authorized")
+    }
```
You can use `o.AssertCallerisOwner`

JJOptimist replied:
I've updated the code to use `Ownable.AssertCallerIsOwner()` instead of the manual check. It uses the built-in assertion, thanks for the suggestion

leohhhn commented on examples/gno.land/r/jjoptimist/home/config.gno:
```diff
@@ -0,0 +1,35 @@
+package home
+
+import (
+    "std"
+    "gno.land/p/demo/ownable"
```
Nitpick-` stdlibs`,`p/` & `r/` should be spaced out:

```suggestion
    "std"
    
    "gno.land/p/demo/ownable"
```

JJOptimist replied:
Noted, i will be separating them from now on.

leohhhn commented on examples/gno.land/r/jjoptimist/home/config.gno:
```diff
@@ -0,0 +1,35 @@
+    config.github = newGithub
+}
+
+func GetOwner() std.Address {
+    return o.Owner()
+} 
```
Redundant if you export `Ownable`

JJOptimist replied:
I've removed the redundant `GetOwner()` function since I am now exporting `Ownable`. Users can directly call `Ownable.Owner()` to get the owner's address, making the wrapper function unnecessary.

leohhhn commented on examples/gno.land/r/jjoptimist/home/config.gno:
```diff
@@ -0,0 +1,35 @@
+    github:      "jjoptimist",
+}
+
+var o = ownable.NewWithAddress(std.Address("g16vfw3r7zuz43fhky3xfsuc2hdv9tnhvlkyn0nj"))
```
```suggestion
var Ownable = ownable.NewWithAddress(std.Address("g16vfw3r7zuz43fhky3xfsuc2hdv9tnhvlkyn0nj"))
```

`Ownable` should be exported. By exporting a struct such as `ownable`, you also export all of it's methods, which are required to do anything with the owner, such as transfer ownership or drop it. With your implementation it's impossible to update the owner :)

Look into "safe objects" (search the examples folder for this), and read into MsgCall and MsgRun

JJOptimist replied:
Thanks for explaining this! I've exported `Ownable` as suggested. I've checked MsgRun and how cross-realm method calling works

leohhhn commented on examples/gno.land/r/jjoptimist/home/home.gno:
````diff
@@ -0,0 +1,80 @@
+
+func getGnomeArt(height int64) string {
+    var art string
+    if height%7 == 0 {
+        art = gnomeArt4  // winking gnome
+    } else if height%5 == 0 {
+        art = gnomeArt3  // starry-eyed gnome
+    } else if height%3 == 0 {
+        art = gnomeArt2  // happy gnome
+    } else if height%2 == 0 {
+        art = gnomeArt1  // regular gnome
+    } else {
+        art = gnomeArt1  // default
+    }
+    return "```\n" + art + "\n```\n"
````
Nit: check how this would look with a switch

JJOptimist replied:
Changed to switch and looks cleaner now

leohhhn commented on examples/gno.land/r/jjoptimist/home/config.gno:
```diff
@@ -0,0 +1,35 @@
+
+var o = ownable.NewWithAddress(std.Address("g16vfw3r7zuz43fhky3xfsuc2hdv9tnhvlkyn0nj"))
+
+func GetConfig() struct { title string; description string; github string } {
+    return config
+}
```
This is just ugly; define this `config` struct properly.

JJOptimist replied:
Updated , much cleaner now

leohhhn commented on examples/gno.land/r/jjoptimist/home/home.gno:
```diff
@@ -0,0 +1,80 @@
+}
+
+func Render(path string) string {
+    cfg := GetConfig()
```
Since this is the same package, just reference the existing `config` object.

JJOptimist replied:
Updated, I'm using `config` directly now

leohhhn reviewed (changes requested):
Good start; please fix the things I mentioned in the comments below :)

leohhhn commented on examples/gno.land/r/jjoptimist/home/home.gno:
````diff
@@ -0,0 +1,80 @@
+    return "```\n" + art + "\n```\n"
+}
+
+func init() {
+    creation = time.Now()
+}
````
Also register to the hall of fame :)

JJOptimist replied:
Registered ✔

JJOptimist commented:
I've updated the home realm:

- Changed gnome art variables to constants for better code organization
- Replaced if-else chain with switch statement for cleaner gnome art selection
- Properly defined Config struct with exported fields for better type safety
- Exported Ownable and implemented AssertCallerIsOwner for proper access control
- Removed redundant GetOwner() function since Ownable is now exported
- Changed `require` to `dependencies` in gno.mod following new standards
- Properly spaced out imports for better readability
- Registered realm in Hall of Fame for community visibility
- Removed ownership tests to avoid testing imported package functionality
- Using direct config access in same package for better efficiency

leohhhn commented on examples/gno.land/r/jjoptimist/home/config.gno:
```diff
@@ -0,0 +1,33 @@
+    Description: "Exploring Gno and building on-chain",
+    Github:      "jjoptimist",
+}
+
+
```
```suggestion

```

leohhhn commented on examples/gno.land/r/jjoptimist/home/home.gno:
```diff
@@ -0,0 +1,85 @@
+ -"-`
+)
+
+var (
+    creation time.Time
+)
```
```suggestion
var creation time.Time
```

leohhhn commented on examples/gno.land/r/jjoptimist/home/home.gno:
````diff
@@ -0,0 +1,85 @@
+    creation time.Time
+)
+
+func getGnomeArt(height int64) string {
+    var art string
+    switch {
+    case height%7 == 0:
+        art = gnomeArt4  // winking gnome
+    case height%5 == 0:
+        art = gnomeArt3  // starry-eyed gnome
+    case height%3 == 0:
+        art = gnomeArt2  // happy gnome
+    default:
+        art = gnomeArt1  // regular gnome
+    }
+    return "```\n" + art + "\n```\n"
+}
````
Move this to the bottom or somewhere else, since this is a helper function

JJOptimist replied:
Done in [this commit](https://github.com/user/repo/commit/b2e6f03) again after revert.

leohhhn commented on examples/gno.land/r/jjoptimist/home/config.gno:
```diff
@@ -0,0 +1,31 @@
+    Github      string
+}
+
+var config = Config {
+    Title:       "JJOptimist's Home Realm 🏠",
+    Description: "Exploring Gno and building on-chain",
+    Github:      "jjoptimist",
+}
+var Ownable = ownable.NewWithAddress(std.Address("g16vfw3r7zuz43fhky3xfsuc2hdv9tnhvlkyn0nj"))
```
```suggestion
var (
	config = Config {
	    Title:       "JJOptimist's Home Realm 🏠",
	    Description: "Exploring Gno and building on-chain",
	    Github:      "jjoptimist",
	}
	Ownable = ownable.NewWithAddress(std.Address("g16vfw3r7zuz43fhky3xfsuc2hdv9tnhvlkyn0nj"))
)
```

leohhhn commented on examples/gno.land/r/jjoptimist/home/home.gno:
```diff
@@ -0,0 +1,83 @@
+    "strconv"
+    "time"
+
+    "gno.land/p/demo/ownable"
```
This is unused

JJOptimist replied:
Removed the one causing test to fail in [commit](https://github.com/gnolang/gno/pull/3405/commits/d674a91057228cf738f635f0f7c8c012fe10e501)

leohhhn reviewed (approved):
Just check the leftover comments, and run `make fmt` in examples :)

leohhhn commented:
Hey it seems you have a bit of an issue with the diff. Can you just have the files you're trying to add?

JJOptimist commented:
> Hey it seems you have a bit of an issue with the diff. Can you just have the files you're trying to add? After that we can merge.

Removed from cache👍

leohhhn commented on examples/gno.land/r/jjoptimist/home/config.gno:
```diff
@@ -0,0 +1,31 @@
+	Title:       "JJOptimist's Home Realm 🏠",
+	Description: "Exploring Gno and building on-chain",
+	Github:      "jjoptimist",
+}
+var Ownable = ownable.NewWithAddress(std.Address("g16vfw3r7zuz43fhky3xfsuc2hdv9tnhvlkyn0nj"))
+
```
```suggestion
}

var Ownable = ownable.NewWithAddress(std.Address("g16vfw3r7zuz43fhky3xfsuc2hdv9tnhvlkyn0nj"))

```
